[0.21] Rate limit the processing of rumoured addresses

sipa opened the pull request:
Backport of #22387.

The rate at which IP addresses are rumoured (through ADDR and ADDRV2 messages) on the network seems to vary from 0 for some non-participating nodes, to 0.005-0.025 addr/s for recent Bitcoin Core nodes. However, the current codebase will happily accept and process an effectively unbounded rate from attackers. There are measures to limit the influence attackers can have on the addrman database (bucket restrictions based on source IPs), but still - there is no need to permit them to feed us addresses at a rate that's orders of magnitude larger than what is common on the network today, especially as it will cause us to spam our peers too.
    
This PR implements a [token bucket](https://en.wikipedia.org/wiki/Token_bucket) based rate limiter, allowing an average of 0.1 addr/s per connection, with bursts up to 1000 addresses at once. Whitelisted peers as well as responses to GETADDR requests are exempt from the limit. New connections start with 1 token, so as to not interfere with the common practice of peers' self-announcement.

Due to the lack of the `Peer` struct in 0.21, the relevant fields have been added to `CNodeState` instead, necessitating additional locks, and slightly different structure to avoid too much `cs_main` grabbing. The last test-improving commit has also been dropped, as the code has changed too much. Most of the behavior is still tested however, just not the part that compares with RPC statistics.

sipa committed:
Functional tests for addr rate limiting

Github-Pull: #22387
Rebased-From: b4ece8a1cda69cc268d39d21bba59c51fa2fb9ed

sipa committed:
Add logging and addr rate limiting statistics

Includes logging improvements by Vasil Dimov and John Newbery.

Github-Pull: #22387
Rebased-From: f424d601e1b6870e20bc60f5ccba36d2e210377b

sipa committed:
Randomize the order of addr processing

Github-Pull: #22387
Rebased-From: 5648138f5949013331c017c740646e2f4013bc24

jnewbery commented on src/net_processing.cpp:
```diff
@@ -387,6 +394,16 @@ struct CNodeState {
+    /** When m_addr_token_bucket was last updated */
+    std::chrono::microseconds m_addr_token_timestamp{GetTime<std::chrono::microseconds>()};
+    /** Total number of addresses that were dropped due to rate limiting. */
+    std::atomic<uint64_t> m_addr_rate_limited{0};
+    /** Total number of addresses that were processed (excludes rate limited ones). */
+    std::atomic<uint64_t> m_addr_processed{0};
```
These don't strictly need to be atomic since they're guarded by cs_main now, but there's no harm in leaving them like this.

sipa replied:
Agree, I may address this if I retouch.

jnewbery replied:
Yeah, no need to change the branch. Just pointing it out in case it confused other reviewers.

jnewbery reviewed (commented):
utACK 55fcbde5a9fc35b187a8cb31004e5ce94bdbbedd

vasild reviewed (approved):
ACK 55fcbde5a9fc35b187a8cb31004e5ce94bdbbedd

jnewbery commented:
reACK 651216e70b0cd1aaa1318db07e1d89d9cbb65cf3

sipa commented:
@sipsorcery @fanquake Any idea what is causing the appveyor failure here?

sipsorcery commented:
I can fix the compiler error by adjusting `misc.cpp` line 390 to the below. I'm not enough of a C++ expert to say why.

````
  if (request.context.Has<NodeContext>()) {
        auto & chainClients = request.context.Get<NodeContext>().chain_clients;
        for (const auto& chain_client : chainClients) {
            chain_client->setMockTime(time);
        }
    }
````

sipa committed:
Avoid Appveyor compilation failure

vasild reviewed (approved):
ACK 33332768c7842a665679f4bcccc4504c542f09e6

I don't understand how this:
```diff
-        for (const auto& chain_client : request.context.Get<NodeContext>().chain_clients) {
+        const auto& chain_clients = request.context.Get<NodeContext>().chain_clients;
+        for (const auto& chain_client : chain_clients) {
```
could fix this:
```
C:\projects\bitcoin\src\rpc\misc.cpp(396,1): error C2355: 'this': can only be referenced inside non-static member functions or non-static data member initializers
```

MarcoFalke commented:
> I don't understand how this: ...

Probably a msvc bug in C++11 mode?

sipsorcery commented:
> Probably a msvc bug in C++11 mode?

The msvc build config uses C++14. As an experiment I did try with msvc & C++17 and it generated the same error.

MarcoFalke commented:
Looks like the bug started happening unrelated to any of our changes between May 31st and June 19th (https://github.com/bitcoin/bitcoin/commit/926f76cb205ce9ef085f4945cdb439746b140414)

sipa commented:
Hmm, now a fuzzer job times out?

jnewbery commented:
reACK 33332768c7

achow101 commented:
ACK 33332768c7842a665679f4bcccc4504c542f09e6

MarcoFalke commented:
I don't understand why this needs the cs_main lock. The following backport compiled just fine:

```diff
diff --git a/src/net_permissions.h b/src/net_permissions.h
index bba0ea1695..3b841ab138 100644
--- a/src/net_permissions.h
+++ b/src/net_permissions.h
@@ -32,3 +32,4 @@ enum NetPermissionFlags {
     PF_MEMPOOL = (1U << 5),
-    // Can request addrs without hitting a privacy-preserving cache
+    // Can request addrs without hitting a privacy-preserving cache, and send us
+    // unlimited amounts of addrs.
     PF_ADDR = (1U << 7),
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index f29be8d8a3..019d00e7d6 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -148,2 +148,9 @@ static constexpr uint32_t MAX_GETCFHEADERS_SIZE = 2000;
 static constexpr size_t MAX_PCT_ADDR_TO_SEND = 23;
+/** The maximum rate of address records we're willing to process on average. Can be bypassed using
+ *  the NetPermissionFlags::Addr permission. */
+static constexpr double MAX_ADDR_RATE_PER_SECOND{0.1};
+/** The soft limit of the address processing token bucket (the regular MAX_ADDR_RATE_PER_SECOND
+ *  based increments won't go above this, but the MAX_ADDR_TO_SEND increment following GETADDR
+ *  is exempt from this limit. */
+static constexpr size_t MAX_ADDR_PROCESSING_TOKEN_BUCKET{MAX_ADDR_TO_SEND};
 
@@ -456,2 +463,8 @@ struct Peer {
 
+    /** Number of addr messages that can be processed from this peer. Start at 1 to
+     *  permit self-announcement. */
+    double m_addr_token_bucket{1.0};
+    /** When m_addr_token_bucket was last updated */
+    std::chrono::microseconds m_addr_token_timestamp{GetTime<std::chrono::microseconds>()};
+
     Peer(NodeId id) : m_id(id) {}
@@ -2440,2 +2453,5 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             pfrom.fGetAddr = true;
+            // When requesting a getaddr, accept an additional MAX_ADDR_TO_SEND addresses in response
+            // (bypassing the MAX_ADDR_PROCESSING_TOKEN_BUCKET limit).
+            peer->m_addr_token_bucket += MAX_ADDR_TO_SEND;
         }
@@ -2593,2 +2609,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         int64_t nSince = nNow - 10 * 60;
+
+        // Update/increment addr rate limiting bucket.
+        const auto current_time = GetTime<std::chrono::microseconds>();
+        if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
+            // Don't increment bucket if it's already full
+            const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, std::chrono::microseconds{0});
+            const double increment = std::chrono::duration<double>(time_diff).count() * MAX_ADDR_RATE_PER_SECOND;
+            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_PROCESSING_TOKEN_BUCKET);
+        }
+        peer->m_addr_token_timestamp = current_time;
+
+        const bool rate_limited = !pfrom.HasPermission(NetPermissionFlags::PF_ADDR);
         for (CAddress& addr : vAddr)
@@ -2598,2 +2626,7 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
 
+            // Apply rate limiting.
+            if (rate_limited) {
+                if (peer->m_addr_token_bucket < 1.0) break;
+                peer->m_addr_token_bucket -= 1.0;
+            }
             // We only bother storing full nodes, though this may include

MarcoFalke commented:
merging this is also fine, if you don't want to retouch. The only difference should be unneeded cs_main locks

sipa commented:
@MarcoFalke Is that a backport? Peer doesn't exist in 0.21.

MarcoFalke commented:
https://github.com/bitcoin/bitcoin/blob/0.21/src/net_processing.cpp#L438

sipa committed:
Rate limit the processing of incoming addr messages

While limitations on the influence of attackers on addrman already
exist (affected buckets are restricted to a subset based on incoming
IP / network group), there is no reason to permit them to let them
feed us addresses at more than a multiple of the normal network
rate.

This commit introduces a "token bucket" rate limiter for the
processing of addresses in incoming ADDR and ADDRV2 messages.
Every connection gets an associated token bucket. Processing an
address in an ADDR or ADDRV2 message from non-whitelisted peers
consumes a token from the bucket. If the bucket is empty, the
address is ignored (it is not forwarded or processed). The token
counter increases at a rate of 0.1 tokens per second, and will
accrue up to a maximum of 1000 tokens (the maximum we accept in a
single ADDR or ADDRV2). When a GETADDR is sent to a peer, it
immediately gets 1000 additional tokens, as we actively desire many
addresses from such peers (this may temporarily cause the token
count to exceed 1000).

The rate limit of 0.1 addr/s was chosen based on observation of
honest nodes on the network. Activity in general from most nodes
is either 0, or up to a maximum around 0.025 addr/s for recent
Bitcoin Core nodes. A few (self-identified, through subver) crawler
nodes occasionally exceed 0.1 addr/s.

Github-Pull: #22387
Rebased-From: 0d64b8f709b4655d8702f810d4876cd8d96ded82

sipa commented:
@MarcoFalke It would appear that I need better glasses. Fixed.

achow101 commented:
ACK 2a5710805195ca54a02aff3540ceaefb9cb3b3e2

GeneFerneau commented:
Approach + code review ACK [2a57108](https://github.com/bitcoin/bitcoin/pull/22569/commits/2a5710805195ca54a02aff3540ceaefb9cb3b3e2)

jnewbery commented:
reACK 2a5710805195ca54a02aff3540ceaefb9cb3b3e2
